python/lsst/ctrl/bps/quantum_clustering_funcs.py

```diff
@@ -50,57 +50,174 @@ def single_quantum_clustering(config, qgraph, name):
+
+            # Currently getQuantaForTask is currently a mapping taskDef to
+            # Quanta, so quick enough to call repeatedly.
+            task_def = qgraph.findTaskDefByLabel(task_label)
+            quantum_nodes = qgraph.getNodesForTask(task_def)
+
+            # Determine cluster for each node
+            for qnode in quantum_nodes:
+                # Gather info for cluster name template into a dictionary.
+                info = {}
+
+                data_id_info = qnode.quantum.dataId.byName()
+                for dim_name in cluster_dims:
+                    _LOG.debug("dim_name = %s", dim_name)
+                    if dim_name in data_id_info:
+                        info[dim_name] = data_id_info[dim_name]
+                if "equalDimensions" in cluster_config[cluster_label]:
+                    equal_dims = cluster_config[cluster_label]["equalDimensions"]
+                    for pair in [pt.strip() for pt in equal_dims.split(",")]:
+                        dim1, dim2 = pair.strip().split(":")
+                        if dim1 in cluster_dims and dim2 in data_id_info:
+                            info[dim1] = data_id_info[dim2]
+                        elif dim2 in cluster_dims and dim1 in data_id_info:
+                            info[dim2] = data_id_info[dim1]
+
+                info["label"] = cluster_label
+                _LOG.debug("info for template = %s", info)
+
+                # Use dictionary plus template format string to create name.
+                # To avoid # key errors from generic patterns, use defaultdict.
+                cluster_name = template.format_map(defaultdict(lambda: "", info))
+                cluster_name = re.sub("_+", "_", cluster_name)
+
+                # Some dimensions contain slash which must be replaced.
+                cluster_name = re.sub("/", "_", cluster_name)
+                _LOG.debug("cluster_name = %s", cluster_name)
+
+                # Save mapping for use when creating dependencies.
+                quantum_to_cluster[qnode.nodeId] = cluster_name
+
+                # Add cluster to the ClusteredQuantumGraph.
+                # Saving NodeId instead of number because QuantumGraph API
+                # requires it for creating per-job QuantumGraphs.
+                if cluster_name in cqgraph:
+                    cluster = cqgraph.get_cluster(cluster_name)
+                    assert isinstance(cluster, QuantaCluster)
+                else:
+                    cluster = QuantaCluster(cluster_name, cluster_label, info)
+                    cqgraph.add_cluster(cluster)
+                cluster.add_quantum(qnode.nodeId, task_label)
+
+    # Assume any task not handled above is supposed to be 1 cluster = 1 quantum
+    for task_def in qgraph.iterTaskGraph():
+        if task_def.label not in task_labels_seen:
+            _LOG.info("Creating 1-quantum clusters for task %s", task_def.label)
+            task_labels_seen.add(task_label)
```

Comment:
Shouldn't it read `task_label_seen.add(task_def.label)`?

Reply:
iterTaskGraph shouldn't actually ever repeat a task_def.  Removing line completely.